autovalet_system/src/autovalet_system.py

```diff
@@ -0,0 +1,248 @@
+        self.goal_pub           = rospy.Publisher(self.goal_topic,PoseStamped,queue_size=1)
+        self.goalGenerator      = goal_generator(self.map_frame)
+        self.current_goal       = PoseStamped()
+
+        # parking setup ###########################################################
+        # tag_topic - name of the topic in which the pose of the marker is being published
+        # husky_frame is the name of the husky's frame we want to keep track of
+        # aruco_frame_name is the NAME OF THE FRAME PUBLISHED AS THE ARUCO TAG
+        self.tag_topic          = '/ARUCO/pose'
+        self.husky_frame        = 'base_link'
+        self.aruco_frame_name   = 'parking_spot' #'aruco_marker_frame' or 'parking_spot'
+                            # (Needs to be same as what is set in aruco launcher)
+        
+        self.parker = parking_spot(self.goal_topic,
+                                   self.tag_topic,
+                                   self.map_frame,
+                                   self.husky_frame,
+                                   self.aruco_frame_name,
+                                   debug=False)
+
+    # helper fxn to load the correct lane detection params and initialize LaneDetector class
+    def init_detector(self, colorInfo_topic, laneCloud_topic, egoLine_topic):
+        yaml_path = rospkg.RosPack().get_path('autovalet_lane_detection')
+        yaml_path += "/config/"
+        if self.sim:
+            yaml_path += "sim_params.yaml"
+        else:
+            yaml_path += "hardware_params.yaml"
+        
+        f = open(yaml_path,'r')
+        yaml_file = yaml.load(f)
+        f.close()
+        rosparam.upload_params('/lane_detection/',yaml_file)
+
+        hlsBounds = rospy.get_param('/lane_detection/hlsbounds')
+        lineParams = rospy.get_param('/lane_detection/lineparams')
+
+        LD = LaneDetector(colorInfo_topic, 
+                          laneCloud_topic, 
+                          egoLine_topic, 
+                          hlsBounds, 
+                          lineParams,
+                          debug=False)
+
+        self.ld_init = True
+
+        return LD
+
+    # callback for image messages. this is what keeps the system moving forward, as processState gets
+    # called everytime we get a new img
+    def registered_image_callback(self, color_msg, depth_msg):
+        # cvBridge image
+        self.color_img = self.bridge.imgmsg_to_cv2(color_msg, desired_encoding="bgr8")
+        self.depth_img = self.bridge.imgmsg_to_cv2(depth_msg, desired_encoding="passthrough")
+
+        self.depth_frame_id = depth_msg.header.frame_id
+        
+        # if we're not in the PARK state AND the lane detector has been successfully initialized, detect the lane and publish
+        if self.current_state != State.PARK and self.ld_init:
+            # lane detection algo
+            _, self.ego_line = self.laneDetector.detectLaneRGBD(self.color_img, self.depth_img)
+
+        self.processState()
```

Comment:
Use the main while loop to process the state instead of the callback. new method ?